Add function estimating Function bounds over a BoundingBox

simonsticko opened the pull request:
Add a function taylor_estimate_function_bounds, which estimates the
range of a Function over a BoundingBox using a Taylor expansion. Both
the range of the function value and each gradient component is
estimated.

Add another function, lower_bound_on_abs, which converts the lower and
upper bounds to a lower bound on the absolute value of the function,
C \leq |f|.

simonsticko commented:
Need this as part of implementing something else. I'm not really sure if this is too specific to be in deal.II/base or not.

Rombur reviewed (commented):
I have a few minor comments. This needs a changelog. TBH I don't really like the name of the functions but I don't have anything better.

simonsticko commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * $L \leq f(x) \leq U$.
+ */
+struct FunctionBounds
```
Not really. I just think `max`/`min` are much better names than `first`/`second`.
Would you prefer `std::pair`?

Rombur replied:
I don't have a strong preference. I am just wondering what other people think.

bangerth replied:
I think a `std::pair` is good enough.

bangerth replied:
In C++17, you will be able to write
```
  double [min,max] = taylor_estimate_function_bounds(...);
```
to decompose a pair right away.

simonsticko commented:
Fixed the comments, except for the discussion on `std::pair` vs using a custom struct.
Waiting on more feedback before I change it.

bangerth commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+
+/**
+ * Estimate bounds on the value and each gradient component of a Function, $f$,
+ * over a BoundingBox, by approximating it as a 2nd order Taylor polynomial
```
```suggestion
 * over a BoundingBox, by approximating it by a 2nd order Taylor polynomial
```

bangerth commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+ * $F = \sum_i |\partial_i f(x_c)| h_i
+ *    + 1/2 \sum_i \sum_j |\partial_i \partial_j f(x_c)| h_i h_j$.
+ *
+ * Here, $h_i$ is half the side length of the box, which is the distance we
```
```suggestion
 * Here, $h_i$ is half the side length of the box in the $i$th coordinate direction, which is the distance we
```

bangerth commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+ */
+template <int dim>
+void
+taylor_estimate_function_bounds(
```
would it make sense to put this function into `namespace Functions` or a new namespace `FunctionTools`?

simonsticko replied:
If `FunctionTools` is better, do you think the header should be `function_tools.h` instead of `function_bounds.h`?

bangerth replied:
Yes, probably. We may want to add more functions to the namespace.

bangerth commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+ * by estimating it from the incoming lower and upper bounds:
+ * $L \leq f(x) \leq U$.
+ *
+ * By rewriting the above as
```
what does "the above" refer to? The documentation of the other function? I think you probably want to be explicit here.

bangerth commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+ * $L \leq f(x) \leq U$.
+ *
+ * By rewriting the above as
+ * $F - C \leq f(x) \leq F + C \LeftRightArrow |f(x) - F| \leq C$,
```
It is not clear to me how this function can determine F and C from just its argument. Can you make that connection?

bangerth commented on include/deal.II/base/function_bounds.h:
```diff
@@ -0,0 +1,112 @@
+ * negative.
+ */
+double
+lower_bound_on_abs(const FunctionBounds &function_bounds);
```
On second thought about this function, I think it is too specialized. I can see the point of the function above, but would suggest that you keep this one where you actually need it. I think that would also address the issues with the documentation: The function needs substantially more space for documentation that for the implementation because one has to explain the context, and if you just inlined the computations this function does into the place where you currently use it, the context is easier to understand for a reader.

simonsticko commented:
Alright:
 - Changed so that the bounds are returned as a `std::pair<double, double>`
 - Moved the function to a new namespace `FunctionTools`
 - Renamed the file to `function_tools.h`
 - Removed the function `lower_bound_on_abs`.

bangerth commented:
/rebuild
